### DISCUSSION TITLE
Welcome to zebra-mac-label-automator Discussions!

### DISCUSSION THREAD

**Discussion by john-stephens** [john-stephens/zebra-mac-label-automator#1] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to get feedback from the Mac Zebra thermal printer community. We hope that you:
  * Ask questions you’re wondering about.
  * Let us know if the Automator workflow is working for you or not.

**Comment by 007vivek007** [comment]
@john-stephens Thanks a ton for working on this issue mate! I highly appreciate your efforts. However, I'm getting "Failed to determine the printer media!" error (screenshot attached). Here's the output of lpstat -p | awk '{print $2}' | xargs -t -n1 -I {} lpoptions -p {} -l :
"lpoptions -p Zebra_Technologies_ZTC_ZD230_203dpi_ZPL -l
PageSize/Media Size: w90h18 w90h162 w108h18 w108h36 w108h72 w108h144 w144h26 w144h36 w144h72 w144h90 w144h288 w144h396 w162h36 w162h90 w162h288 w162h396 w171h396 w180h72 w180h144 w198h90 w216h72 w216h90 w216h144 w216h216 *w216h360 w234h144 w234h360 w234h396 w234h419 w234h563 w252h72 w288h72 w288h144 w288h180 w288h216 w288h288 w288h360 w288h432 w288h468 w288h936 w432h72 w432h144 w432h216 w432h288 w432h360 w432h432 w432h468 w576h72 w576h144 w576h216 w576h288 w576h360 w576h432 w576h468 Custom.WIDTHxHEIGHT
Resolution/Resolution: *203dpi 300dpi 600dpi
zeMediaTracking/Media Tracking: *Continuous Web Mark
MediaType/Media Type: Saved Thermal *Direct
Darkness/Darkness: -1 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 *18 19 20 21 22 23 24 25 26 27 28 29 30
zePrintRate/Print Rate: Default *1 2 3 4 5 6 7 8 9 10 11 12
zeLabelTop/Label Top: *200 -120 -115 -110 -105 -100 -95 -90 -85 -80 -75 -70 -65 -60 -55 -50 -45 -40 -35 -30 -25 -20 -15 -10 -5 0 5 10 15 20 25 30 35 40 45 50 55 60 65 70 75 80 85 90 95 100 105 110 115 120
zePrintMode/Print Mode: *Saved Tear Peel Rewind Applicator Cutter
zeTearOffPosition/Tear-Off Adjust Position: *1000 -120 -115 -110 -105 -100 -95 -90 -85 -80 -75 -70 -65 -60 -55 -50 -45 -40 -35 -30 -25 -20 -15 -10 -5 0 5 10 15 20 25 30 35 40 45 50 55 60 65 70 75 80 85 90 95 100 105 110 115 120
zeErrorReprint/Reprint After Error: *Saved Always Never"

Please let me know how to get rid of this error.
![Screenshot 2024-11-07 at 11 14 43 PM](https://github.com/user-attachments/assets/169d0293-b2b9-4834-9cb6-01d51fc2e5e8)

**Reply by john-stephens** [reply]
Thanks for reaching out! I'm guessing that the size of the document being printed doesn't exactly match one of the Zebra Media Size options. I'll put out a new version later that provides some debug into in the dialog. I may need to add some logic to round to the nearest media size.

**Reply by john-stephens** [reply]
Just published v1.1.1. Give that a try and let me know what the error message says this time.

**Reply by 007vivek007** [reply]
Thanks for the prompt attention to this matter. It's giving me the same error message i.e. Failed to determine the Print Media!. Attaching the error screenshot with the message here.
![Screenshot 2024-11-08 at 8 20 42 AM](https://github.com/user-attachments/assets/1abcf727-eef9-4629-a9f5-6f7af891b721)

**Reply by john-stephens** [reply]
Thanks for the updated screenshot! That's helpful. I'll dig into it.

**Reply by john-stephens** [reply]
I suspect that you have a different version of `bc` than I do. On the command line, can you run `bc --version` and provide the output here? The output of `which bc` would also be helpful.

**Reply by 007vivek007** [reply]
Mine is 1.06 (Screenshot attached)
![Screenshot 2024-11-08 at 6 18 37 PM](https://github.com/user-attachments/assets/cd6b4168-cb4d-48a3-9380-29bf769ef57c)

**Reply by john-stephens** [reply]
Which version of MacOS do you have? Can you also provide the output of `which bc`?

**Reply by 007vivek007** [reply]
I'm on macOS Monterey 12.7.6 (21H1320). The output of `which bc` is "/usr/bin/bc".

**Reply by john-stephens** [reply]
Check out latest release. I think I have this fixed. The main issue is that MacOS 12 has an older version of `bc` compared to what I am using in MacOS 14 and doesn't support the `-e` option. Let me know!

**Reply by 007vivek007** [reply]
Highly appreciate your hard work but it's still giving me media error. Attaching the updated screenshot below:
![Screenshot 2024-11-08 at 10 34 23 PM](https://github.com/user-attachments/assets/6a5dd1e3-1976-49fb-99bc-4e07353cc43f)

**Reply by john-stephens** [reply]
Should be fixed now. Grab the latest again. Appreciate you trying this out!

**Reply by 007vivek007** [reply]
Finally got the workflow to start :-) Thanks, but it messed up the barcodes on the label. The bars are too dark and clubbed together, making the barcode unscannable. I've attached images of the original label and the output below. Please let me know if you can fix this.
![Untitled-1](https://github.com/user-attachments/assets/979cfd6e-c2ea-4114-8f11-d35f873eeceb)

This output is with darkness 5 setting. I have tried it with Darkness 1 as well but the result is same i.e. unscannable.

**Reply by john-stephens** [reply]
Would you be able to provide a copy of the full label? I'd like to try it out and see if I can adjust the thresholding settings.

**Reply by 007vivek007** [reply]
Sure, the full label can be downloaded from [here](https://we.tl/t-X685QJ9ZZa)

**Reply by john-stephens** [reply]
Thanks. Confirmed that I see the same thing on my side. Let me work on this.

**Reply by john-stephens** [reply]
Try the latest release. I adjusted the threshold and it worked a ton better on my side.

**Reply by 007vivek007** [reply]
Yay! Finally got the crisp quality. You're a life saviour 😍 Thanks a ton man for fixing the threshold issue. You rocked ❤️‍🔥 BUT the only issue is when I'm giving print command for multiple pages pdf file, it gives me same "Failed to determine printer media" error. However, it works like a charm for a single page pdf file.

**Reply by 007vivek007** [reply]
The only issue is when I'm giving print command for multiple pages pdf file, it gives me same "Failed to determine printer media" error. However, it works like a charm for a single page pdf file (Attaching the updated screenshot below)
![Screenshot 2024-11-09 at 1 35 39 AM](https://github.com/user-attachments/assets/cd258541-6e3a-4ad4-a573-b308ef7cb90b)

**Reply by john-stephens** [reply]
Would you be able to provide a copy of the multipage PDF? The functionality is theoretically there, but obviously there is a bug. Thanks again for testing this out!

**Reply by 007vivek007** [reply]
Please download the multi pages pdf from [here](https://we.tl/t-xAce5ZDPvI)

**Reply by john-stephens** [reply]
Should be fixed. Grab the latest.

**Reply by 007vivek007** [reply]
The multipage issue was fixed. Thanks. But there are two issues:

1. When executing the multipage print command, there is a 3-5 seconds delay between the pages. So, if I'm printing a 100-page PDF, I'll have to wait 5 secondsx100, or 8 minutes. Is this normal?
2. The printed layout area on a label isn't static. It's moving from its original space in a multipage file. On the fifth or sixth page, it's printing out of the page, i.e., on the previous or next label. So, in the end, one label is printed on two pages. (Please refer to the screenshot [here](https://we.tl/t-56DYiLxTFB))

**Reply by john-stephens** [reply]
Regarding the first point, it currently prints one page at a time. That would likely explain the delay. I'll have to see if I can convert all of the pages and then spool all of them to the printer to speed it up. Let me think about that.

Regarding the second point, I suspect that you may need to calibrate your printer. Refer to the manual and give that a shot. If you are still having issues, let me know.

**Reply by john-stephens** [reply]
Give the latest version a try. Rather than trying to print each page at a time, all of the pages are converted into a single file which is then printed. At least for my printer, it seems to print 2 pages at a time, pauses briefly, then continues. I suspect that is a limitation of my printer. Let me know how it works for you.

**Reply by 007vivek007** [reply]
Thanks a ton man for fixing the delay issue. Now it prints all pages in one go. 

Regarding calibration: I attempted to calibrate the printer by pressing the top button until the green light flashed twice. While the calibration was successful, the original issue persists. The printing layout continues to shift off the paper gradually with each page. By the sixth or seventh page, the layout started cutting off the barcodes

**Reply by john-stephens** [reply]
I'm honestly not sure. Assuming that the printer is calibrated and the media size selected in the print dialog is the same as the media in the printer, there shouldn't be any issues. Does it work correctly when you just print normally without running the Automator workflow?

**Reply by 007vivek007** [reply]
> I'm honestly not sure. Assuming that the printer is calibrated and the media size selected in the print dialog is the same as the media in the printer, there shouldn't be any issues. Does it work correctly when you just print normally without running the Automator workflow?

I fixed it by changing the media type from "continuous" to "non-continuous". Thanks a lot for your help and efforts.

**Reply by 007vivek007** [reply]
Hi John! Thank you for the fantastic Automator you developed—it's working well for vertical labels! However, I'm still encountering issues with the horizontal labels. I have uploaded both files [here](https://file.io/XvSC7w77bZwQ), which include the original label file and the output file.

Could you please take a look? Out of the five labels, the worst and unscannable ones were from Shadowfax and Valmo logistics partners.

Thanks for your help!

**Reply by 007vivek007** [reply]
Any updates here? @john-stephens

**Reply by john-stephens** [reply]
I'm not sure on this. I printed all of the labels at once and they all came out perfectly on my machine. I'd recommend checking the manual for your printer. You could also try adjusting your darkness. Mine is currently set to 16.

**Reply by 007vivek007** [reply]
My darkness setting was set to 14 when I sent you the output sample. I'm unable to scan the label even with Darkness setting 6. I'm not sure what to do now.

> I'm not sure on this. I printed all of the labels at once and they all came out perfectly on my machine. I'd recommend checking the manual for your printer. You could also try adjusting your darkness. Mine is currently set to 16.

**Comment by gitty8** [comment]
iam using custome labels, 50x80m and had same issue: failed to determine the printer media!
i fixed the issue with some modification and setting in cups setting the default media size to 50x80mm.

```
PRINTER="Zebra"
IDENTIFY="/opt/homebrew/bin/identify"
MAGICK="/opt/homebrew/bin/magick"
PDFTOPPM="/opt/homebrew/bin/pdftoppm"
PDFINFO="/opt/homebrew/bin/pdfinfo"

# Check that the specified printer exists
if ! lpstat -p "$PRINTER"; then
  osascript -e "display dialog \"Printer \\\"$PRINTER\\\" not found.\n\nPlease check the \$PRINTER setting.\""
  exit
fi

# Check that the required apps are installed
for cmd in "$IDENTIFY" "$MAGICK" "$PDFTOPPM" "$PDFINFO"; do
  if ! command -v "$cmd" > /dev/null 2>&1; then
    osascript -e "display dialog \"\\\"$cmd\\\" not found.\n\nPlease verify installation.\""
    exit
  fi
done

# Set the media size manually
media="w50h80"
dpi="$(lpoptions -p "$PRINTER" -l | grep "^Resolution" | sed -E "s/^.*\*([0-9]+)dpi.*/\1/g")"

if [[ ! "$dpi" =~ ^[0-9]+$ ]]; then
  osascript -e "display dialog \"Failed to determine the printer DPI!\""
  exit
fi

for f in "$@"; do
  tmpdir=$(mktemp -d)
  if [[ "$?" -gt 0 ]]; then
    osascript -e "display dialog \"Failed to create temp dir!\""
    exit
  fi

  # Extract all of the pages
  $PDFTOPPM -r $dpi -aa no -aaVector no -gray "$f" $tmpdir/page

  # Apply thresholding to all pages and spool to the printer
  $MAGICK $tmpdir/page* -colorspace gray -color-threshold 'gray(50%)-gray(100%)' -density $dpi -compress none pdf:- | \
  lpr -P "$PRINTER" -o portrait -o media="$media"

  # Cleanup
  rm -rf "$tmpdir"
done
```

**Reply by john-stephens** [reply]
The Automator workflow tries to determine the print media based on the size of the PDF that is given to it. ~~Just curious... was your print size 50x80mm? If that's the case and it didn't work for you, can you provide the output of `lpstat -p | awk '{print $2}' | xargs -t -n1 -I {} lpoptions -p {} -l` and a copy of the PDF you are trying to print?~~ I don't think there is a direct default equivalent to 50x80mm in the Zebra CUPS driver. The equivalent would be something like "w142h227" (convert to inches and multiply by 72). It sounds like you added a custom media size, is that correct?

**Reply by john-stephens** [reply]
@gitty8 Give the latest version a try. It adds support for custom media sizes.

**Comment by petetum** [comment]
works on GK420d on mb air m2
Thank you!

**Reply by john-stephens** [reply]
Thanks for the feedback! I added it to the list.
